Database functions

ConchobarMacNessa opened the pull request:
I did a few things in this PR:
1. tidied up white space in db_build.sql
2. Added the dummy data from the gist to the db, and linked to this in d3.js (rather than calling to the gist)
3. Wrote a function (and a route) to get the data from the database (in db_get.js)
4. Wrote two functions to post to the database. The first pushes a new memory to the memories and media tables. The second pushes a new user to the database for when we implement sign up
5. Wrote a has password function to use in the db call mentioned above. 

relates #40 #41 #48

alice-carr committed:
sorted out whitespace
cleaned up whitespace in sql file
relates #38

alice-carr committed:
updated schema to fit our data
relates #40

alice-carr committed:
added dummy data
added dummy data and sorted out spacing
relates #38 #40

alice-carr committed:
updated db
with dummy data and sorted out spacing
relates #38 #40

alice-carr committed:
added get memories function
relates #41

alice-carr committed:
merging with master

alice-carr committed:
set up home route
and get data routes in express server
relates #41

alice-carr committed:
linked d3.js in index
and made call to route in d3.js
relates #41

alice-carr committed:
set up route for db get func
so now we can pass data to our d3.js
relates #41

alice-carr committed:
slight refactor to keep eslint happy
relates #41

alice-carr committed:
added db post func
to post new memory to db
relates #41

alice-carr committed:
updated sql build
so that there is a starting value in the array
relates #41

alice-carr committed:
updated all media columns in dbbuild
so they are initialised with empty curly brackets
this way we can push to them
relates #41;

alice-carr committed:
added hashpassword js
added function to hash password and installed bcrypt
relates #48

alice-carr committed:
added register user post func
set up so that when we implement sign up
relates #48 #41

alice-carr committed:
removed test route
I was previouslt using a test route to test posting a new memory
this has now been removed
relates #41

des-des commented on database_build/db_build.sql:
```diff
@@ -3,29 +3,52 @@ BEGIN;
-  audio    TEXT[],
-  video    TEXT[],
-  image    TEXT[]
+  text_only   INTEGER[],
```
Why us this an array of integers?

ConchobarMacNessa replied:
I made this an array of integers so that we could store the ids of the memories that correspond to each media type – e.g. if a memory is a video, we will store its id in the video array. 

When we initially talked last week you mentioned database 'normalisation' I think, is this not the best way to go about it?

des-des replied:
Is sql, you describe relations between tables using foreign keys.

In this case, the relation is `belongs to`.

Ie a `memory_image` belongs to a `memory`.

So, you will have something like
```sql
CREATE TABLE memory_text (
    text varchar(100) NOT NULL,
    FOREIGN KEY (memory_id) REFERENCES memories(memory_id)
);
```
Not super hot on my sql so that might need tweeking, but thats the general idea

des-des commented on database_build/db_build.sql:
```diff
@@ -3,29 +3,52 @@ BEGIN;
-  video    TEXT[],
-  image    TEXT[]
+  text_only   INTEGER[],
+  audio       INTEGER[],
```
and this

des-des commented on database_build/db_build.sql:
```diff
@@ -3,29 +3,52 @@ BEGIN;
-  image    TEXT[]
+  text_only   INTEGER[],
+  audio       INTEGER[],
+  video       INTEGER[],
```
and this

des-des commented on database_build/db_build.sql:
```diff
@@ -3,29 +3,52 @@ BEGIN;
+  text_only   INTEGER[],
+  audio       INTEGER[],
+  video       INTEGER[],
+  image       INTEGER[]
```
and this

des-des commented on src/database/db_get.js:
```diff
@@ -0,0 +1,14 @@
+const get = {};
+
+get.memories = (userId, callback) => {
+  connect.query('SELECT memories.id, memories.heading, memories.likes, memories.avgRating, memories.visits, memories.tag, memories.memory FROM memories INNER JOIN users ON users.id = memories.user_id WHERE users.id = $1', [userId], (err, res) => {
```
Can you pull this string out into a multiline variable and add indentation + new lines

des-des commented on src/database/db_get.js:
```diff
@@ -0,0 +1,14 @@
+  });
+};
+
+module.exports = get;
```
I think stick to export style like:
```js
module.exports = {
  memories
}
```
at the bottom of each file. That is the convention I used when I browserified the client side code so should be consistent

des-des commented on src/database/db_post.js:
```diff
@@ -0,0 +1,44 @@
+
+post.memory = (newMemory, callback) => {
+  connect.query(
+    `WITH new_memory AS (
```
Yeah, I can read a query written like this!

des-des commented on src/database/db_post.js:
```diff
@@ -0,0 +1,44 @@
+      RETURNING id
+    )
+    UPDATE media SET
+      ${newMemory.mediaType} = ${newMemory.mediaType} || (SELECT id FROM new_memory)`,
```
Okay, you should not inject values directly into the query like this. It leaves you vulnerable to **sql injection attacks**

des-des commented on src/database/db_post.js:
```diff
@@ -0,0 +1,44 @@
+    });
+};
+
+module.exports = post;
```
Yeah, same, keep export style consistent, I do not have a prefference, but do your exports the same thoughout codebase (see client side code in big d3 pull request)

des-des commented on src/database/db_post.js:
```diff
@@ -0,0 +1,44 @@
+const connect = require('./db_connect');
```
With the file name `db_post`, I am not convinced grouping queries by http method is the best way.

Rather than
`post.memory()`, `post.user()`, `get.memory()` ...

I would structure stuff like
`memory.create(data)`, `memory.get(memoryId)`

In terms of the verbs:

HTTP METHOD | function name
---|---
GET | get
POST | create
PUT | Update
DELETE | delete

Does that make sense?

des-des commented on src/routes/getD3Data.js:
```diff
@@ -0,0 +1,18 @@
+const app = express.Router();
+
+module.exports = [
+  app.get('/getd3data', (req, result) => {
```
This is not a good endpoint name

better:

`GET /memories?userId={userId}`

You do not have to use the userId, but your endpoints keep the same format thought development

ConchobarMacNessa replied:
Is it ok for us to leave this as `GET /memories` ?

alice-carr committed:
spaced out sql query in db get
relates #41

alice-carr committed:
renamed functions
relates #41

alice-carr committed:
removed sql injection
relates #41

alice-carr committed:
changed route url
relates #41

alice-carr committed:
added sql query to delete memory
relates #20

alice-carr committed:
merged with master

alice-carr committed:
starting to link up db to d3.js
relates #41

alice-carr committed:
merging master with db-push
relates #41

alice-carr committed:
changed avgRating to avgrating
as the db was saving as avgrating
relates #41

alice-carr committed:
refactored generateId so that it would only  use none used ids...
this will probably change when we use UUID
relates #41

alice-carr committed:
added conditional to check if running on local host or if it is a local file
relates #41

alice-carr committed:
also updated gist so that it works when the server is not running
relates #41

alice-carr committed:
updated schema
so that media type is now an enum
changed this in the sql queries too
relates #41

alice-carr committed:
updated gist to reflect new schema system
also deleted unnec line of code in d3 js
relates #41

ghost commented:
@des-des I changed the schema as discussed
I also made a call to the db in d3.js so the info we are getting from there is from the db (apart from when we're not running the server – in that case we're still fetching the data from a gist...)

alice-carr committed:
removed incorrect serial key
relates #41
